### PR TITLE
#1495 Combine template and configuration files into template set files

### PR DESCRIPTION
**Pull request by KHucklenbroich**
Adresses/Fixes #1558 #1495.

Check https://github.com/devonfw/cobigen/pull/1558
#1558 
Implements

Requests made until 24.10.2022 are taken into account.

* 
* 
* 

@devonfw/cobigen

### CODE REVIEW

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/reader/ContextConfigurationReader.java [hunk 1]
```diff
@@ -66,6 +93,122 @@ public ContextConfigurationReader(Path configRoot) throws InvalidConfigurationEx
...
+
+    super(contextRoot);
+
+    if (this.contextConfigurations == null) {
+      this.contextConfigurations = new HashMap<>();
+    }
+
+    this.contextRoot = contextRoot;
+    this.contextConfigurations.put(contextRoot, contextConfiguration);
+    // TODO: check if this is still needed
```
Comment (EduardKrieger) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/reader/ContextConfigurationReader.java:114]: open TODO

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/entity/Matcher.java [hunk 2]
```diff
@@ -22,13 +37,13 @@ public class Matcher extends AbstractMatcher {
    * @param type matcher type
    * @param value to match against
    * @param variableAssignments of the matcher
-   * @param accumulationType of the matcher
+   * @param accumulationType2 of the matcher
```
Comment (EduardKrieger) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/entity/Matcher.java:40]: Change back to accumulationType

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/entity/Matcher.java [hunk 3]
```diff
@@ -4,16 +4,31 @@
...
-
 /** The {@link Matcher} type corresponds to the &lt;matcher&gt; xml node */
 public class Matcher extends AbstractMatcher {
 
   /** {@link VariableAssignment}s */
   private List<VariableAssignment> variableAssignments;
 
   /** Accumulation type */
-  private AccumulationType accumulationType;
+  private String accumulationType;
```
Comment (EduardKrieger) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/entity/Matcher.java:14]: Change back to accumulationType

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/constant/TemplateSetConfigurationVersion.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+import java.util.Map;
+
+import com.google.common.collect.Maps;
+
+/** Version steps of the templates configuration. */
+public enum TemplateSetConfigurationVersion {
+
+  /**
+   * Initial release.
```
Comment (EduardKrieger) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/constant/TemplateSetConfigurationVersion.java:11]: Maybe also say that this a templateSet merged the old configurations together

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/constant/TemplateSetConfigurationVersion.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  /**
+   * @return whether the configuration is backward compatible to the previous version.
+   */
+  public boolean isBackwardCompatible() {
+
+    return this.backwardCompatible;
+  }
+
+  /**
+   * Get latest context configuration version supported by this CobiGen release.
```
Comment (EduardKrieger) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/constant/TemplateSetConfigurationVersion.java:50]: template set

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/reader/TemplateSetConfigurationReader.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+  /** The top-level folder where the templates are located. */
+  private TemplateFolder rootTemplateFolder;
+
+  /** The list of adapted template set configuration paths */
+  private List<Path> templateSetConfigurationPathsAdapted;
+
+  /** The list of downloaded template set configuration paths */
+  private List<Path> templateSetConfigurationPathsDownloaded;
+
+  // TODO: Use dependency injection here instead of the new operator
```
Comment (EduardKrieger) [inline cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/reader/TemplateSetConfigurationReader.java:101]: open TODO

**Review by EduardKrieger — CHANGES_REQUESTED** [review]

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/reader/AbstractContextConfigurationReader.java [hunk 7]
```diff
@@ -1,47 +1,19 @@
...
-import com.devonfw.cobigen.impl.config.versioning.VersionValidator;
-import com.devonfw.cobigen.impl.config.versioning.VersionValidator.Type;
-
-import jakarta.xml.bind.JAXBContext;
-import jakarta.xml.bind.JAXBException;
-import jakarta.xml.bind.UnmarshalException;
-import jakarta.xml.bind.Unmarshaller;
 
 /** The {@link ContextConfigurationReader} reads the context xml */
 public abstract class AbstractContextConfigurationReader {
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/reader/AbstractContextConfigurationReader.java:15]: Good question, I think it is not needed anymore and can be moved into the ContextConfigurationReader.

File: cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/constant/TemplateSetConfigurationVersion.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+import com.google.common.collect.Maps;
+
+/** Version steps of the templates configuration. */
+public enum TemplateSetConfigurationVersion {
+
+  /**
+   * Initial release.
+   */
+  v6_0(1f, false);
```
[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/constant/TemplateSetConfigurationVersion.java:13]: I think we should change this version to v1_0 instead. This is the first template-set version and it does not make sense to initialize it with v6_0.

[the remark this answers is not in the document]
Comment (jan-vcapgemini) [reply cobigen/cobigen-core/src/main/java/com/devonfw/cobigen/impl/config/constant/TemplateSetConfigurationVersion.java:13]: The only "problem" I see is that we need to change this version in all template-set.xmls then...

**Pull request merged**